xDiffCollection/CollectionBin.swift

```diff
@@ -0,0 +1,86 @@
+//
+//  CollectionBin.swift
+//  CollectionTest
+//
+//  Created by Esteban Garro on 2018-06-05.
+//  Copyright © 2018 aldo. All rights reserved.
+//
+
+import Foundation
+
+internal enum BinOperationType {
+    case updated
+    case added
+    case deleted
+    case none
+}
+
+internal struct BinResult {
+    var type: BinOperationType
+    var idx : Int
+}
+
+internal protocol CollectionFiltering {
+    func filter<T:Hashable>(_ element:T) -> Bool
+}
+
+internal struct CollectionBin<T:Hashable> : CollectionFiltering, CustomStringConvertible {
+    var index : Int
+    var elements:Array<T> = []
+    var filter:(T)->Bool
+    var description: String {
+        var resp = ""
+        for (idx, element) in self.elements.enumerated() {
+            resp += "\(idx))\t\(element.hashValue)\n"
+        }
+        return resp
+    }
+    var count : Int {
+        return elements.count
+    }
+    
+    init(index: Int, filter: @escaping (T)->Bool, elements: Array<T>) {
+        self.index = index
+        self.filter = filter
+        self.elements = elements
+    }
+    
+    func filter<T>(_ element:T) -> Bool {
+        return self.filter(element)
+    }
+    
+    func updated(_ element:T) -> (CollectionBin<T>, BinResult) {
+        var copy = elements
+        if let idx = self.contains(element) {
```

Comment:
Want to try with reduce?

Reply:
Hey! I have a great idea. Let's keep it like that, since all the tests pass. And for our next functional programming session, we can use this as perfect examples for practice!! What do you think?!

Reply:
Well, actually not a bad idea 👍 
I don't mind if the team is ok. @beljamali @piyush9290  what do you think guys?

Reply:
I agree, we'll practice having this as an example. Looks like Choice 1 has Rank 1 😛

Reply:
agree for that!